kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java

```diff
@@ -499,348 +532,174 @@ protected void handleInactive(KafkaHeaderAndRequest kafkaHeaderAndRequest,
-                                        });
-                            } else {
-                                completeOne.run();
-                            }
-                        });
             }
-        }).exceptionally(error -> {
-            topicMapFuture.completeExceptionally(error);
-            return null;
         });
-        return topicMapFuture;
+        return topicAndMetadataList;
+    }
+
+    private CompletableFuture<List<String>> authorizeNamespacesAsync(final Collection<String> namespaces,
+                                                                     final AclOperation aclOperation) {
+        final Map<String, CompletableFuture<Boolean>> futureMap = namespaces.stream().collect(
+                Collectors.toMap(
+                        namespace -> namespace,
+                        namespace -> authorize(aclOperation, Resource.of(ResourceType.NAMESPACE, namespace))
+                ));
+        return CoreUtils.waitForAll(futureMap.values()).thenApply(__ ->
+            futureMap.entrySet().stream().filter(e -> {
+                if (!e.getValue().join()) {
+                    log.warn("Failed to authorize {} for ACL operation {}", e.getKey(), aclOperation);
+                    return false;
+                }
+                return true;
+            }).map(Map.Entry::getKey).collect(Collectors.toList())
+        );
+    }
+
+    private CompletableFuture<Stream<String>> listAllTopicsFromNamespacesAsync(final List<String> namespaces) {
+        return CoreUtils.waitForAll(namespaces.stream()
+                .map(namespace -> pulsarService.getNamespaceService()
+                        .getListOfPersistentTopics(NamespaceName.get(namespace))
+                ).collect(Collectors.toList()),
+                topics -> topics.stream().flatMap(List::stream));
+    }
+
+    private CompletableFuture<ListPair<String>> authorizeTopicsAsync(final Collection<String> topics,
+                                                                     final AclOperation aclOperation) {
+        final Map<String, CompletableFuture<Boolean>> futureMap = topics.stream().collect(
+                Collectors.toMap(
+                        topic -> topic,
+                        topic -> authorize(aclOperation, Resource.of(ResourceType.TOPIC, topic))
+                ));
+        return CoreUtils.waitForAll(futureMap.values()).thenApply(__ ->
+                ListPair.of(futureMap.entrySet()
+                        .stream()
+                        .collect(Collectors.groupingBy(e -> e.getValue().join()))
+                ).map(Map.Entry::getKey));
+    }
+
+    private CompletableFuture<List<TopicAndMetadata>> findTopicMetadata(final ListPair<String> listPair,
+                                                                        final boolean allowTopicAutoCreation) {
+        final Map<String, CompletableFuture<Integer>> futureMap = CoreUtils.listToMap(
+                listPair.getSuccessfulList(),
```

Comment:
what happens to topics with errors that are not AUTHORIZATION_FAILURE  ?

it seems to me that below we are reporting them all with AUTHORIZATION_FAILURE

Reply:
> we are reporting them all with AUTHORIZATION_FAILURE

No. Only those topics are failed to be authorized are reported with AUTHORIZATION_FAILURE.

```java
        ).thenApply(topicAndMetadataList -> { // [1]
            final List<TopicAndMetadata> list = CoreUtils.listToList(listPair.getFailedList(),
                    topic -> new TopicAndMetadata(topic, TopicAndMetadata.AUTHORIZATION_FAILURE)); // [2]
            list.addAll(topicAndMetadataList);
            return list;
        });
```
- `[1]` (`topicAndMetadataList`) is the lookup result by `getPartitionedTopicMetadataAsync`, the error (`numPartitions` field) could also be `INVALID_PARTITIONS`.
- `[2]` (the initial `list`) contains the failed topics from `authorizeTopicsAsync` so they are all reported with `AUTHORIZATION_FAILURE`.

Since `[1]` could be an empty list `Collections.emptyList()` so we cannot call `addAll` on `[1]`. Instead, we should create a new list and call `addAll`, which accepts `[1]` as the argument.

See `getTopicsAsync`:

```java
            return authorizeTopicsAsync(fullTopicNames, AclOperation.DESCRIBE)
                    .thenCompose(authorizedTopicsPair -> findTopicMetadata(authorizedTopicsPair, allowTopicCreation));
```

Reply:
I'll rename them to make the code more readable.

Reply:
See https://github.com/streamnative/kop/pull/1107/commits/9424ea9a4b7016629e27d5f695a997df2d951364.